### DISCUSSION TITLE
`NumberInput.value` vs `int`

### DISCUSSION THREAD

**Discussion by maratori** [lona-web-org/lona#102] [Feature Request]
In Lona 1.7 [`NumberInput`](https://github.com/lona-web-org/lona/blob/03106a884012e8a286e4e6e5d1f6048ac1d95701/lona/html/data_binding/inputs.py#L144) was added to work with `<input type="number">`. It was undocumented because we may want to change it in the next release.

Now all properties (`value`, `step`, `min`, `max`) have `Optional[float]` type.
```py
node = NumberInput(value=10, step=2)
assert node.value == 10
assert node.step == 2
assert type(node.value) is float
assert type(node.step) is float
```

Most users will work with whole numbers only. And they will have to convert `int(node.value)` each time. It seems reasonable to provide some builtin option to have `int` directly.

I did two failed attempts in #87. The first was to add separate `IntNumberInput`, the second is to determine type based on `step`. Both seem bad options. Why? In two words value inside browser can be `float` even if it is invalid (see [details](https://github.com/lona-web-org/lona/pull/87#issuecomment-919393332)).

Let's add new read-only property `NumberInput.int_value`.
* It will raise exception if non-integers are valid (for example `NumberInput(step=0.1).int_value`)
* It will return `int` if current `value` is valid integer
* It will return `None` if current `value` is invalid (for example `NumberInput(value=3, min=5).int_value is None` and `NumberInput(value=0.5, min=0, step=1).int_value is None`)

If you see better options, you are welcome.

**Comment by fscherf** [comment]
@maratori I did some research on the browser side

- Even though the type of the input is ``number`` the value of ``input.value`` is always a string
- You can enter, and therefore send, invalid values, but the return value of ``input.checkValidity()`` is ``false`` then

**NumberInput.value**

I think it would be most consistent to copy the behavior of the browser at let ``NumberInput.value`` be a string. For easy handling we should add ``to_int()`` and ``to_float()``. 

To have ``NumberInput.value`` return ``int`` by default would be very convenient though.
On html5 level there seems no way to distinguish between ``int`` and ``float``. If we want that, we would have to reconsider a split into ``NumberInput`` and ``FloatNumberInput`` (or something like that) or a keyword for ``NumberInput`` like ``base_type=int``.

**NumberInput.valid**

I am not completely sure about this property. We could let the browser send the return code of ``input.checkValidity()`` for each CHANGE event, but that only works for one direction and not when ``NumberInput.value`` is set by the view. Maybe we should add checks in the python code and ignore what the browser thinks is valid (?)

**Reply by maratori** [reply]
> Even though the type of the input is `number` the value of `input.value` is always a string

There is also `input.valueAsNumber` in browser - the same as current `node.value`.

> I think it would be most consistent to copy the behavior of the browser at let `NumberInput.value` be a string.

I don't think we should copy browser in all aspects. I can't imagine why we may need to get raw string.
But if it is necessary, it's better to add dedicated property returning `str`.

> On html5 level there seems no way to distinguish between `int` and `float`. If we want that, we would have to reconsider a split into `NumberInput` and `FloatNumberInput` (or something like that) or a keyword for `NumberInput` like `base_type=int`.

I've already tried to implement that in #87. The problem described above and in https://github.com/lona-web-org/lona/pull/87#issuecomment-919393332. How `base_type=int` should behave if value is invalid?

> NumberInput.valid

Good point. I think we should add this property.
And it will be just few lines of Python code. No need to get it from browser.
Algorithm is described here https://github.com/lona-web-org/lona/pull/87#issuecomment-919393332

**Reply by fscherf** [reply]
@maratori

> I don't think we should copy browser in all aspects. I can't imagine why we may need to get raw string. But if it is necessary, it's better to add dedicated property returning str.
 
I think you are right. ``NumberInput.raw_value`` maybe? This should be an unsetable property.

> How `base_type=int` should behave if value is invalid?

The input event contains the value as a string. The ``NumberInput`` should try ``self.base_type(value)`` and if this fails, ``value`` should be set to the invalid string. ``NumberInput.valid`` should be``False`` in this case.

> > NumberInput.valid
> 
> Good point. I think we should add this property.
> And it will be just few lines of Python code. No need to get it from browser.
> Algorithm is described here [#87 (comment)](https://github.com/lona-web-org/lona/pull/87#issuecomment-919393332)

OK :+1:

Btw: We could run [reportValidity()](https://googlechrome.github.io/samples/report-validity/) on the frontend before sending the input event.

**Reply by maratori** [reply]
> `NumberInput.raw_value` maybe? This should be an unsetable property.

👍 

> `value` should be set to the invalid string.

What means "invalid string"? For example we have `node = NumberInput(base_type=int)`. User types `1.5` in browser. What value should be after that `node.value == ???`

> Btw: We could run [reportValidity()](https://googlechrome.github.io/samples/report-validity/) on the frontend before sending the input event.

It is not always good. Real-world example.
I created an app with [dash](https://github.com/plotly/dash) to account my stocks.
Dividends are in cents (2 digits after point) most of the time.
So it would be great to set `step=0.01` to be able to change value with up/down keys.
But I can't do it because dash doesn't allow to send invalid number to server.
From the other side some dividends are in cents fraction (for example 4 digits after point).
And it would be invalid number.

**Reply by fscherf** [reply]
> What means "invalid string"? For example we have `node = NumberInput(base_type=int)`. User types `1.5` in browser. What value should be after that `node.value == ???`

In this case it would make sense if the value was ``'1.5'`` because it comes in as a string and can't be converted lossless. 

> > Btw: We could run [reportValidity()](https://googlechrome.github.io/samples/report-validity/) on the frontend before sending the input event.
> 
> It is not always good. Real-world example.
> I created an app with [dash](https://github.com/plotly/dash) to account my stocks.
> Dividends are in cents (2 digits after point) most of the time.
> So it would be great to set `step=0.01` to be able to change value with up/down keys.
> But I can't do it because dash doesn't allow to send invalid number to server.
> From the other side some dividends are in cents fraction (for example 4 digits after point).
> And it would be invalid number.

Thats why i proposed to use only the report call. I agree that invalid data should be sent, so the view can show a custom error message in these cases. [reportValidity()](https://googlechrome.github.io/samples/report-validity/) only shows a browser specific popup that the value is invalid. That behavior should be configurable though. We could add a property ``report_validity`` which sets ``data-lona-report-validity`` on the node element.

**Reply by maratori** [reply]
> In this case it would make sense if the value was '1.5' because it comes in as a string and can't be converted lossless.

If I understood you correctly, `NumberInput.value` will have following type `Union[int, float, str]`.
It goes against my original idea to have fixed type to avoid conversion in user code.

**Reply by fscherf** [reply]
Hm. You are right. What about this:

```python
class NumberInput:
    def handle_input_event(self, input_event):
        self.raw_value = input_event.data

        try:
            self.value = self.base_class(self.raw_value)
            self.valid = self.check_validity()  # check step, min, max etc

        except ValueError:
            self.valid = False
            self.value = None
```

**Reply by fscherf** [reply]
@maratori does this collide with your idea or use cases?

**Reply by maratori** [reply]
@fscherf Looks good.
Am I right you want to have `base_class = int` by default?
If yes here are questions:
1. Following `NumberInput(min=0.5)` will never have `.valid==True`. Is it ok?
2. What `base_class` will be after parsing html string? Always `int`?

**Reply by fscherf** [reply]
@maratori Interesting question. We could parse the type of the initial value.

```python
class NumberInput:
    def __init__(self, value=None, base_type=None):
        if value and not base_type:
            self.base_type = type(value)

        elif not base_type:
            self.base_type = int

        else:
            self.base_type = base_type

    def handle_input_event(self, input_event):
        self.raw_value = input_event.data

        try:
            self.value = self.base_type(self.raw_value)
            self.valid = self.check_validity()  # check step, min, max etc

        except ValueError:
            self.valid = False
            self.value = None
```

**Reply by maratori** [reply]
I'm not sure it's good to be so "smart".
Why not just have following properties?
* `NumberInput.raw_value -> str`
* `NumberInput.value -> Optional[float]`
* `NumberInput.int_value -> Optional[int]`
* `NumberInput.valid -> bool`

**Reply by fscherf** [reply]
@maratori I think you are right. Also it makes more sense for ``value`` to be a float than an int the longer i think about it. Last question: Should it be ``int_value`` or ``to_int()`` ?

**Reply by maratori** [reply]
I'm ok with both

**Reply by fscherf** [reply]
Do we need this in the first place? ``number.int_value`` is a one-liner, but ``int(number.int)`` too. You have to check ``valid`` before hand anyway, because ``number.int_value`` can be none

**Reply by maratori** [reply]
That's a great question. I'm not sure (some time ago :) Exactly because it can be `None`.
That's why I'm not pushing this idea anymore.

**Reply by fscherf** [reply]
:) ok, then let's just cut ``int_value`` for now, and agree on

```
NumberInput.raw_value -> str
NumberInput.value -> Optional[float]
NumberInput.valid -> bool
```
